src/stores/walletStore/walletStore.ts

```diff
@@ -37,54 +84,95 @@ export class WalletStore {
     isWalletConnected: false,
   };
   private listeners: Set<Listener> = new Set();
+  private providers: Map<string, EIP6963ProviderDetail> = new Map();
 
   constructor() {
-    const onChainChange = () => {
-      this.connectMetamask();
+    this.setupEIP6963Listeners();
+    this.setupChangeListeners();
+  }
+
+  private setupEIP6963Listeners() {
+    if (typeof window === 'undefined') return;
+
+    window.addEventListener('eip6963:announceProvider', (event: Event) => {
+      const providerEvent = event as EIP6963AnnounceProviderEvent;
+      this.providers.set(providerEvent.detail.info.uuid, providerEvent.detail);
+    });
+
+    window.dispatchEvent(new Event('eip6963:requestProvider'));
+  }
+
+  private setupChangeListeners() {
+    const onChainChange = async () => {
+      await this.refreshCurrentConnection();
     };
 
-    const onAccountChange = () => {
-      this.connectMetamask();
+    const onAccountChange = async () => {
+      await this.refreshCurrentConnection();
     };
 
     this.subscribe(() => {
       const connection: WalletConnection = this.getSnapshot();
-      // @ts-expect-error window.ethereum.off does exist
-      window.ethereum.off('chainChanged', onChainChange);
-      // @ts-expect-error window.ethereum.off does exist
-      window.ethereum.off('accountsChanged', onAccountChange);
-
-      if (connection.walletType === WalletTypes.METAMASK) {
-        // @ts-expect-error window.ethereum.on does exist
-        window.ethereum.on('chainChanged', onChainChange);
-        // @ts-expect-error window.ethereum.on does exist
-        window.ethereum.on('accountsChanged', onAccountChange);
+
+      if (connection.provider) {
+        try {
+          if (connection.provider.off) {
+            connection.provider.off('chainChanged', onChainChange);
+            connection.provider.off('accountsChanged', onAccountChange);
+            connection.provider.off('disconnect', onAccountChange);
+          }
+        } catch (e) {
+          console.warn('Failed to remove listeners', e);
+        }
+      }
+
+      if (
+        connection.isWalletConnected &&
+        connection.provider &&
+        connection.provider.on
+      ) {
+        try {
+          connection.provider.on('chainChanged', onChainChange);
+          connection.provider.on('accountsChanged', onAccountChange);
+          connection.provider.on('disconnect', onAccountChange);
+        } catch (e) {
+          console.warn('Failed to add listeners', e);
+        }
       }
     });
   }
 
+  public getProviders(): EIP6963ProviderDetail[] {
+    window.dispatchEvent(new Event('eip6963:requestProvider'));
+
+    return Array.from(this.providers.values());
+  }
```

Comment:
Raison detre: collect all of the EIP 6963 compatible providers to be used to build a UI to select which one the user wants.